programs/stake/src/stake_state.rs

```diff
@@ -842,6 +817,83 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
+/// After calling `validate_delegate_amount()`, this struct contains calculated values that are used
+/// by the caller.
+struct ValidatedDelegateInfo {
```

Comment:
```suggestion
/// After calling `validate_delegated_amount()`, this struct contains calculated values that are used
/// by the caller.
struct ValidatedDelegatedInfo {
```

Reply:
Remind why we need this to be a struct instead of just returning u64? More info coming?
Seems like we parse it to pull out the stake_amount immediately in all cases.

Reply:
Yeah, this is another one of those fuzzy areas...

The main role of the function is to validate the stake delegation amount. When I see a fn called `validate_delegated_amount()`, it's not immediately obvious (to me) what a `u64` return value would represent. As an specific example, take the function `fn add(x: u64, y: u64) -> u64`, it is immediately clear what this function (likely) returns.

So in the case of `validate_delegated_amount()`, that function calculates what the stake amount is as part of its validation. I forget where I read it, but someone said (paraphrasing):
> Return all values calculated within a function.
Returning _all_ values is likely an exaggeration (or at least impractical for performance reasons), but returning _useful_ calculated values does make sense, since I need the stake amount after validation and I don't want to repeat the calculation again. I like to put these calculated values into an "info" struct for two reasons:

1. The struct communicates the intent; that its fields are informational/side-band, and computed during fn execution.
2. A struct and its fields have names, so it becomes self-documenting. I don't need to flip back to the function's implementation to know what the value represents, as the field's name tells me so. (Assuming the names are descriptive enough (or if there are doc comments added as well).)
 
This is actually something I've found challenging when reading through new areas of the code base, when functions have return values like `(u64, u64, u64, Vec<(Pubkey, u64)>)` and I try to trace that through the stack.

Since this function is not likely a performance bottleneck, I would not expect a u64 wrapped in a struct to result in performance degradation (in fact the Release build should produce identical ASM ([godbolt link](https://godbolt.org/z/TvGhThMcP)).) 

I guess the tl;dr is that there's no performance impact, and it improves (my) comprehension and intent.

Reply:
> when functions have return values like `(u64, u64, u64, Vec<(Pubkey, u64)>)`

All tuples of size 3+ need to be eradicated from the tree in favor of `structs`

Reply:
I'm with you about structifying tuples generally, but I personally think this struct provides more obfuscation than elucidation. 🤷‍♀️ Not a hill I'll die on; take it or leave it ;)

Reply:
Oh! I forgot one. I wrote `validate_split_amount()` first, which has two fields in its struct. So when I wrote `validate_delegated_amount()`, I decided to use the same pattern since it was doing something similar, even though there was only one field in the struct. (FWIW)